### PR TITLE
Updated test_graph_grad_scaling to use new OptimizerInfo infrastructure

### PR DESCRIPTION
**Pull request by FireACNC**
This PR targets the issue mentioned in #123451 , and solves the specific task to update`test_graph_grad_scaling` in `test/test_cuda.py` to use the new OptimizerInfo infrastructure.

`test_graph_grad_scaling` is moved to a new `TestCase` class called `TestCudaOptims` in order to use `instantiate_device_type_tests`. The test content remained the same. `@onlyCUDA` is applied to the new test; the original use of the wrapper function is also changed to a `@parametrize` decorator for better style.

If we think that this migration is successful, we can delete the original test item under `TestCuda`. Currently it is left untouched to avoid any unexpected issues. 

Local linter passed. 
```
$ lintrunner test/test_cuda.py
ok No lint issues.
```

Local tests passed.
```
> python .\test\test_cuda.py -k test_graph_grad_scaling
Ran 7 tests in 0.458s
OK (skipped = 3)
```

### CODE REVIEW

**Comment by FireACNC** [comment]
The changes need further testing, closed for now.

**Comment by FireACNC** [comment]
@janeyx99 Hi! I'm a first-time contributor and I would appreciate your thoughts on how to proceed from here. 
I have some preliminary changes, and then I noticed that all tests in `test_cuda.py` are executed with `instantiate_parametrized_tests`, which doesn't have `device_cls` specified and will raise an error if applying `@opims`. I thought about isolating the test and using `instantiate_device_type_tests` as suggested, but I don't think that is the proper way of handling it, since in the document these tests are designed to be device-agnostic. 
Any help will be greatly appreciated!

**Comment by janeyx99** [comment]
@FireACNC Thanks for looking into this! Feel free to move these tests into their own class within test_cuda.py where you can apply `instantiate_device_type_tests`. That is okay if it improves coverage for these tests. Where do you see the device-agnostic part? I'm not sure I know which part you're looking at.

**Comment by jayanthd04** [comment]
> @FireACNC Thanks for looking into this! Feel free to move these tests into their own class within test_cuda.py where you can apply `instantiate_device_type_tests`. That is okay if it improves coverage for these tests. Where do you see the device-agnostic part? I'm not sure I know which part you're looking at/if it implies we cannot touch the test_cuda tests.

I am having the same issue, ```test_graph_optims ``` and ```test_graph_scaling_fused_optimizers``` are also under the ```TestCuda``` class which is called with ```instantiate_parametrized_test```. Whenever I try to use the ```@optims``` decorator, an error is thrown that the decorator should be used with ```instantiate_device_type_tests()``` instead. I believe it would be best  to move all the optimizer tests in the ```TestCuda``` class to a separate class and run them using ```instantiate_device_type_tests``` since any of the optimizer tests in the class would encounter the same issue, but I am unsure if that change would disrupt the current testing framework's structure. What would be the best way to proceed? 
Thank you in advance!

**Comment by janeyx99** [comment]
@jayanthd04 feel free to add another TestCase class in test_cuda.py for optimizers in order to be able to use the @optims decorator!

**Comment by FireACNC** [comment]
@janeyx99 Thank you for the reply! I'll try to add another TestCase class as well and implement the new optims structure this week.

> Where do you see the device-agnostic part? I'm not sure I know which part you're looking at/if it implies we cannot touch the test_cuda tests.

I saw this in `common_utils.py` under the definition of `instantiate_parametrized_tests`, where it says `This should be used instead of instantiate_device_type_tests() if the test class contains device-agnostic tests`. I'm wondering if this is the specific reason `instantiate_parametrized_tests` is used in `test_cuda.py`. For now, I will try to use `instantiate_device_type_tests()` for the tests I'm working on in order to apply the optims structure, but if this influences any existing feature please let me know and we can think about how to mitigate it!

**Comment by FireACNC** [comment]
@janeyx99 Hi! I've updated the code and the PR description and would really appreciate some review on it! Please let me know if this looks good or what should I be working on from here.

**Review by janeyx99 — COMMENTED** [review]
Did you forget to delete the test in TestCuda?

**Comment by janeyx99** [comment]
@FireACNC looks like you're missing the deletion part of the task 😛 But you got the right idea--I haven't looked at the actual test too closely--it'd help to see what test it's replacing so ping me when that's ready.

**Comment by FireACNC** [comment]
@janeyx99 Thank you so much for the review! Yes, I've deleted the original test now. Please let me know what you think, thanks!

File: test/test_cuda.py [hunk 1]
```diff
@@ -4148,9 +4098,68 @@ def test_no_triton_on_import(self):
...
 
+class TestCudaOptims(TestCase):
+    # These tests will be instantiate with instantiate_device_type_tests
+    # to apply the new OptimizerInfo structure.
+
+    @onlyCUDA
+    @unittest.skipIf(not TEST_CUDA_GRAPH, "CUDA >= 11.0 or ROCM >= 5.3 required for graphs")
+    @parametrize("foreach, fused", [(False, False), (True, False), (False, True)])
+    @optims(
+        optim_db,
```
Comment (janeyx99) [inline test/test_cuda.py:4109]:
```suggestion
        [optim for optim in optim_db if "foreach" in optim.supported_impls and "fused" in optim.supported_impls],
```
Ah, I realize you'll need this, as passing in fused or foreach flags will error on optims that don't support those.

**Review by janeyx99 — APPROVED** [review]
Looks great! One more change and we'll be on our way to merging.

Approving now, but the conditional will have to be fixed so that CI can be green.

**Comment by FireACNC** [comment]
Thanks and I've applied the change. Which conditional are you referring to? If it's something I can help with, please let me know!

**Comment by janeyx99** [comment]
Rebasing to see if the errors are old.

@pytorchbot rebase

**Comment by FireACNC** [comment]
Looks like there's a failing test due to timeout/authorization issue. It seems unrelated to our changes, maybe we can rerun the specific test?

**Pull request closed without merging**